### ISSUE TITLE
Improve insert performance for partitioned tables

### ISSUE THREAD

**Issue by matriv** [crate/crate#15801]
`primaryKeys` for a partitioned table is set to an empty list here: https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/metadata/doc/DocTableInfo.java#L262 which makes sense for select queries, as `_id` is not unique across the partitions. But this has a side-effect that during inserts we cannot take advantage of the optimized insert path: https://github.com/crate/crate/blob/master/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java#L1035

From a benchmark on a modified `uservisits` table:
```
CREATE TABLE uservisits (
       "sourceIP" STRING,
       "destinationURL" STRING,
       "visitDate" TIMESTAMP,
       "adRevenue" FLOAT,
       "UserAgent" STRING INDEX USING FULLTEXT,
       "cCode" STRING,
       "lCode" STRING,
       "searchWord" STRING,
       "duration" INTEGER,
       "y" GENERATED ALWAYS AS date_trunc('year', "visitDate"),
       INDEX uagent_plain USING PLAIN("UserAgent")
       -- ^^ used for regex matching ^^ --
    ) PARTITIONED BY("y") WITH (
        number_of_replicas = 0
    );
```
using the following spec:
```
[setup]
statement_files = ["sql/uservisits.sql"]
statements = [
    "copy uservisits from 'https://cdn.crate.io/downloads/datasets/amplab/1node/uservisits/part-00001.gz' with (compression = 'gzip')",
    "copy uservisits from 'https://cdn.crate.io/downloads/datasets/amplab/1node/uservisits/part-00002.gz' with (compression = 'gzip')",
    "copy uservisits from 'https://cdn.crate.io/downloads/datasets/amplab/1node/uservisits/part-00003.gz' with (compression = 'gzip')",
    "copy uservisits from 'https://cdn.crate.io/downloads/datasets/amplab/1node/uservisits/part-00004.gz' with (compression = 'gzip')",
    "copy uservisits from 'https://cdn.crate.io/downloads/datasets/amplab/1node/uservisits/part-00005.gz' with (compression = 'gzip')",
    "refresh table uservisits"
]

[[load_data]]
target = "uservisits"
source = "https://cdn.crate.io/downloads/datasets/amplab/1node/uservisits/part-00006.gz"
bulk_size = 100
concurrency = 20


[teardown]
statements = ["drop table if exists uservisits"]
```

The results show a significant improvement, v1: `master` v2: a branch where we enforce the insert to go to the optimized path:

```
Q: insert into uservisits ("adRevenue", "destinationURL", "searchWord", "UserAgent", "duration", "visitDate", "sourceIP", "lCode", "cCode") values ($1, $2, $3, $4, $5, $6, $7, $8, $9)
C: 20
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       13.153 ±   24.252 |      1.001 |      7.306 |     11.608 |    250.076 |
|   V2    |        9.015 ±   14.997 |      0.846 |      4.723 |      7.661 |    233.695 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  37.33%                           -  42.94%   
```

The data distribution for this setup is:
```
cr> select size / 1024 / 1024, num_docs, table_name, partition_ident from sys.shards;
+----------------------------------------+----------+------------+--------------------------+
| ((size / 1024::bigint) / 1024::bigint) | num_docs | table_name | partition_ident          |
+----------------------------------------+----------+------------+--------------------------+
|                                      8 |    27182 | uservisits | 04130                    |
|                                      8 |    27394 | uservisits | 04130                    |
|                                      9 |    27371 | uservisits | 04130                    |
|                                      8 |    27431 | uservisits | 04130                    |
|                                      8 |    26808 | uservisits | 04636c9l6cr30c1g60o30    |
|                                      8 |    27176 | uservisits | 04636c9l6cr30c1g60o30    |
|                                      8 |    27262 | uservisits | 04636c9l6cr30c1g60o30    |
|                                      8 |    27213 | uservisits | 04636c9l6cr30c1g60o30    |
|                                      8 |    27093 | uservisits | 0463ccpg6sp30c1g60o30    |
|                                      8 |    27362 | uservisits | 0463ccpg6sp30c1g60o30    |
|                                      9 |    27154 | uservisits | 0463ccpg6sp30c1g60o30    |
|                                      8 |    27185 | uservisits | 0463ccpg6sp30c1g60o30    |
|                                      9 |    27511 | uservisits | 0463id1m74q38c1g60o30    |
|                                      9 |    27302 | uservisits | 0463id1m74q38c1g60o30    |
|                                     14 |    27119 | uservisits | 0463id1m74q38c1g60o30    |
|                                      9 |    27378 | uservisits | 0463id1m74q38c1g60o30    |
|                                      9 |    27368 | uservisits | 046j2chm68pj0d1g60o30c0  |
|                                      8 |    27031 | uservisits | 046j2chm68pj0d1g60o30c0  |
.... (total 172 shards, largest shard ~14MB)
```

**Comment by matriv** [comment]
Did one more test with the following table (2 partitions/1 shard each)
```
CREATE TABLE uservisits (
       "sourceIP" STRING,
       "destinationURL" STRING,
       "visitDate" TIMESTAMP,
       "adRevenue" FLOAT,
       "UserAgent" STRING INDEX USING FULLTEXT,
       "cCode" STRING,
       "lCode" STRING,
       "searchWord" STRING,
       "duration" INTEGER,
       "y" GENERATED ALWAYS AS date_trunc('year', "visitDate") / 1000000000000,
       INDEX uagent_plain USING PLAIN("UserAgent")
       -- ^^ used for regex matching ^^ --
    ) CLUSTERED INTO 1 shards PARTITIONED BY("y") WITH (
        number_of_replicas = 0
    );
```
loading first the data files 0-8, and benchmarking the 9th.
0-8 preloaded data files yield the following shard info:
```
cr> select size / 1024 / 1024, num_docs, table_name, partition_ident from sys.shards;
+----------------------------------------+----------+------------+-----------------+
| ((size / 1024::bigint) / 1024::bigint) | num_docs | table_name | partition_ident |
+----------------------------------------+----------+------------+-----------------+
|                                    971 |  5259365 | uservisits |           04130 |
|                                    373 |  1699738 | uservisits |           04132 |
+----------------------------------------+----------+------------+-----------------+
```

and the benchmark results are:
```
Q: insert into uservisits ("adRevenue", "destinationURL", "searchWord", "UserAgent", "duration", "visitDate", "sourceIP", "lCode", "cCode") values ($1, $2, $3, $4, $5, $6, $7, $8, $9)
C: 20
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        6.815 ±   14.030 |      1.956 |      4.119 |      4.939 |    182.224 |
|   V2    |        4.156 ±   17.014 |      0.692 |      1.945 |      2.600 |    291.667 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  48.47%                           -  71.69%   
```

**Comment by seut** [comment]
We've implemented this improvement by #15870. It will be available with the next 5.6 and 5.7 hotfix releases.
